File: Data/Entities/DocumentoExpediente.cs
using System.ComponentModel.DataAnnotations;
using SisMortuorio.Data.Entities.Enums;

namespace SisMortuorio.Data.Entities;

/// <summary>
/// Documento digitalizado adjunto al expediente mortuorio.
/// Reemplaza los "juegos de copias físicas" del proceso manual.
/// 
/// FLUJO:
/// 1. Admisionista sube el archivo escaneado (Estado: PendienteVerificacion)
/// 2. Admisionista verifica físicamente contra el original presentado
/// 3. Admisionista marca como verificado (Estado: Verificado)
/// 4. Sistema calcula Expediente.DocumentacionCompleta automáticamente
/// 
/// DOCUMENTOS REQUERIDOS según TipoSalida:
/// - Familiar → DNI_Familiar + DNI_Fallecido + CertificadoDefuncion
/// - AutoridadLegal → OficioLegal
/// </summary>
public class DocumentoExpediente
{
    /// <summary>
    /// Identificador único del documento
    /// </summary>
    [Key]
    public int DocumentoExpedienteID { get; set; }

    // ===================================================================
    // RELACIÓN CON EXPEDIENTE
    // ===================================================================

    /// <summary>
    /// ID del expediente al que pertenece este documento
    /// </summary>
    [Required]
    public int ExpedienteID { get; set; }

    /// <summary>
    /// Navegación al expediente
    /// </summary>
    public virtual Expediente Expediente { get; set; } = null!;

    // ===================================================================
    // CLASIFICACIÓN DEL DOCUMENTO
    // ===================================================================

    /// <summary>
    /// Tipo de documento según catálogo
    /// Determina si es obligatorio según TipoSalida del ActaRetiro
    /// </summary>
    [Required]
    public TipoDocumentoExpediente TipoDocumento { get; set; }

    /// <summary>
    /// Estado de verificación del documento
    /// PendienteVerificacion → Verificado → (o) Rechazado
    /// </summary>
    [Required]
    public EstadoDocumentoExpediente Estado { get; set; } = EstadoDocumentoExpediente.PendienteVerificacion;

    // ===================================================================
    // ARCHIVO
    // ===================================================================

    /// <summary>
    /// Ruta relativa del archivo en el servidor
    /// Formato: "documentos-expedientes/2025/01/SGM-2025-00001_DNI_Familiar_20250115.pdf"
    /// </summary>
    [Required]
    [MaxLength(500)]
    public string RutaArchivo { get; set; } = string.Empty;

    /// <summary>
    /// Nombre original del archivo subido por el usuario
    /// Ejemplo: "dni_juan_perez.jpg"
    /// </summary>
    [Required]
    [MaxLength(255)]
    public string NombreArchivo { get; set; } = string.Empty;

    /// <summary>
    /// Extensión del archivo en minúsculas
    /// Valores permitidos: .pdf, .jpg, .jpeg, .png
    /// </summary>
    [Required]
    [MaxLength(10)]
    public string ExtensionArchivo { get; set; } = string.Empty;

    /// <summary>
    /// Tamaño del archivo en bytes
    /// Máximo permitido: 5 MB (5,242,880 bytes)
    /// </summary>
    [Required]
    public long TamañoBytes { get; set; }

    // ===================================================================
    // AUDITORÍA - SUBIDA
    // ===================================================================

    /// <summary>
    /// ID del usuario que subió el documento
    /// Típicamente el Admisionista
    /// </summary>
    [Required]
    public int UsuarioSubioID { get; set; }

    /// <summary>
    /// Navegación al usuario que subió el documento
    /// </summary>
    public virtual Usuario UsuarioSubio { get; set; } = null!;

    /// <summary>
    /// Fecha y hora en que se subió el documento
    /// </summary>
    [Required]
    public DateTime FechaHoraSubida { get; set; } = DateTime.Now;

    // ===================================================================
    // AUDITORÍA - VERIFICACIÓN
    // ===================================================================

    /// <summary>
    /// ID del usuario que verificó el documento contra el original físico
    /// Típicamente el mismo Admisionista que lo subió
    /// Solo tiene valor cuando Estado == Verificado o Rechazado
    /// </summary>
    public int? UsuarioVerificoID { get; set; }

    /// <summary>
    /// Navegación al usuario que verificó el documento
    /// </summary>
    public virtual Usuario? UsuarioVerifico { get; set; }

    /// <summary>
    /// Fecha y hora en que se verificó o rechazó el documento
    /// </summary>
    public DateTime? FechaHoraVerificacion { get; set; }

    // ===================================================================
    // OBSERVACIONES
    // ===================================================================

    /// <summary>
    /// Observaciones del Admisionista al subir o verificar el documento
    /// Ejemplos:
    /// - "DNI con fecha de vencimiento próxima, se aceptó"
    /// - "Imagen legible, coincide con original presentado"
    /// - "RECHAZADO: archivo borroso, solicitar nuevo escaneo"
    /// </summary>
    [MaxLength(500)]
    public string? Observaciones { get; set; }

    // ===================================================================
    // MÉTODOS DE LÓGICA DE NEGOCIO
    // ===================================================================

    /// <summary>
    /// Obtiene el tamaño del archivo en formato legible
    /// Ejemplo: "1.23 MB", "450 KB"
    /// </summary>
    public string ObtenerTamañoLegible()
    {
        string[] sizes = ["B", "KB", "MB", "GB"];
        double len = TamañoBytes;
        int order = 0;

        while (len >= 1024 && order < sizes.Length - 1)
        {
            order++;
            len /= 1024;
        }

        return $"{len:0.##} {sizes[order]}";
    }

    /// <summary>
    /// Verifica si el documento está verificado y válido
    /// </summary>
    public bool EsValido() => Estado == EstadoDocumentoExpediente.Verificado;

    /// <summary>
    /// Marca el documento como verificado por el admisionista
    /// </summary>
    /// <param name="usuarioID">ID del usuario que verifica</param>
    /// <param name="observaciones">Observaciones opcionales de la verificación</param>
    public void MarcarVerificado(int usuarioID, string? observaciones = null)
    {
        Estado = EstadoDocumentoExpediente.Verificado;
        UsuarioVerificoID = usuarioID;
        FechaHoraVerificacion = DateTime.Now;
        Observaciones = observaciones ?? Observaciones;
    }

    /// <summary>
    /// Marca el documento como rechazado indicando el motivo
    /// El familiar deberá presentar nuevamente el documento
    /// </summary>
    /// <param name="usuarioID">ID del usuario que rechaza</param>
    /// <param name="motivo">Motivo del rechazo (obligatorio para trazabilidad)</param>
    public void MarcarRechazado(int usuarioID, string motivo)
    {
        Estado = EstadoDocumentoExpediente.Rechazado;
        UsuarioVerificoID = usuarioID;
        FechaHoraVerificacion = DateTime.Now;
        Observaciones = $"RECHAZADO: {motivo}";
    }
}

File: sgm-frontend/src/app/services/documento-expediente.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

// ═══════════════════════════════════════════════════════════
// ENUMS
// ═══════════════════════════════════════════════════════════

export enum TipoDocumentoExpediente {
  DNI_Familiar = 1,
  DNI_Fallecido = 2,
  CertificadoDefuncion = 3,
  OficioLegal = 4,
  ActaLevantamiento = 5,
  Otro = 6
}

export enum EstadoDocumentoExpediente {
  PendienteVerificacion = 1,
  Verificado = 2,
  Rechazado = 3
}

// ═══════════════════════════════════════════════════════════
// INTERFACES - RESPUESTA
// ═══════════════════════════════════════════════════════════

/**
 * Documento digitalizado del expediente (respuesta del backend)
 */
export interface DocumentoExpedienteDTO {
  documentoExpedienteID: number;
  expedienteID: number;

  // Clasificación
  tipoDocumento: TipoDocumentoExpediente;
  tipoDocumentoDescripcion: string;
  estado: EstadoDocumentoExpediente | string;
  estadoDescripcion: string;

  // Archivo
  nombreArchivo: string;
  extensionArchivo: string;
  tamanioLegible: string;

  // Auditoría - Subida
  usuarioSubioNombre: string;
  fechaHoraSubida: string;

  // Auditoría - Verificación
  usuarioVerificoNombre?: string;
  fechaHoraVerificacion?: string;

  // Observaciones
  observaciones?: string;
}

/**
 * Estado individual de un tipo de documento requerido.
 * Usado en el semáforo visual del frontend.
 */
export interface EstadoDocumentoItem {
  subido: boolean;
  verificado: boolean;
  rechazado: boolean;
  documentoID?: number;
  nombreArchivo?: string;
  observaciones?: string;
}

/**
 * Resumen del estado de documentación del expediente.
 * Determina si el botón "Crear Acta" está habilitado.
 */
export interface ResumenDocumentosDTO {
  expedienteID: number;

  /** true → habilita botón "Crear Acta" en frontend */
  documentacionCompleta: boolean;

  /** TipoSalida del ActaRetiro si existe, null si aún no se creó */
  tipoSalida?: 'Familiar' | 'AutoridadLegal' | null;

  // Documentos caso Familiar
  dniFamiliar: EstadoDocumentoItem;
  dniFallecido: EstadoDocumentoItem;
  certificadoDefuncion: EstadoDocumentoItem;

  // Documentos caso AutoridadLegal
  oficioLegal: EstadoDocumentoItem;

  // Lista completa
  documentos: DocumentoExpedienteDTO[];
}

// ═══════════════════════════════════════════════════════════
// INTERFACES - REQUESTS
// ═══════════════════════════════════════════════════════════

export interface RechazarDocumentoDTO {
  documentoExpedienteID: number;
  motivo: string;
  usuarioVerificoID: number;
}

// ═══════════════════════════════════════════════════════════
// SERVICIO
// ═══════════════════════════════════════════════════════════

/**
 * Servicio para gestión de documentos digitalizados del expediente.
 * Reemplaza el proceso manual de "3 juegos de copias físicas".
 *
 * Endpoints:
 * - GET  /api/DocumentosExpediente/expediente/{id}          → lista documentos
 * - GET  /api/DocumentosExpediente/expediente/{id}/resumen  → semáforo
 * - GET  /api/DocumentosExpediente/{id}                     → detalle
 * - POST /api/DocumentosExpediente/subir                    → upload archivo
 * - POST /api/DocumentosExpediente/{id}/verificar           → marcar verificado
 * - POST /api/DocumentosExpediente/{id}/rechazar            → marcar rechazado
 * - GET  /api/DocumentosExpediente/{id}/descargar           → descarga
 * - DELETE /api/DocumentosExpediente/{id}                   → eliminar
 */
@Injectable({
  providedIn: 'root'
})
export class DocumentoExpedienteService {
  private http = inject(HttpClient);
  private readonly apiUrl = 'https://localhost:7153/api/DocumentosExpediente';

  // ═══════════════════════════════════════════════════════════
  // CONSULTAS
  // ═══════════════════════════════════════════════════════════

  /**
   * Obtiene todos los documentos de un expediente
   */
  obtenerPorExpediente(expedienteId: number): Observable<DocumentoExpedienteDTO[]> {
    return this.http.get<DocumentoExpedienteDTO[]>(
      `${this.apiUrl}/expediente/${expedienteId}`
    );
  }

  /**
   * Obtiene el resumen de documentación del expediente.
   * Incluye semáforo visual por tipo de documento y DocumentacionCompleta.
   */
  obtenerResumen(expedienteId: number): Observable<ResumenDocumentosDTO> {
    return this.http.get<ResumenDocumentosDTO>(
      `${this.apiUrl}/expediente/${expedienteId}/resumen`
    );
  }

  /**
   * Obtiene un documento por su ID
   */
  obtenerPorId(documentoId: number): Observable<DocumentoExpedienteDTO> {
    return this.http.get<DocumentoExpedienteDTO>(`${this.apiUrl}/${documentoId}`);
  }

  // ═══════════════════════════════════════════════════════════
  // SUBIDA
  // ═══════════════════════════════════════════════════════════

  /**
   * Sube un documento digitalizado al expediente.
   * Usa multipart/form-data — no JSON.
   * Formatos: .pdf, .jpg, .jpeg, .png — Máximo 5MB.
   */
  subirDocumento(
    expedienteId: number,
    tipoDocumento: TipoDocumentoExpediente,
    archivo: File,
    observaciones?: string
  ): Observable<DocumentoExpedienteDTO> {
    const formData = new FormData();
    formData.append('expedienteId', expedienteId.toString());
    formData.append('tipoDocumento', tipoDocumento.toString());
    formData.append('archivo', archivo, archivo.name);
    if (observaciones) {
      formData.append('observaciones', observaciones);
    }

    return this.http.post<DocumentoExpedienteDTO>(
      `${this.apiUrl}/subir`,
      formData
    );
  }

  // ═══════════════════════════════════════════════════════════
  // VERIFICACIÓN / RECHAZO
  // ═══════════════════════════════════════════════════════════

  /**
   * Marca un documento como verificado contra el original físico.
   * Actualiza DocumentacionCompleta del expediente automáticamente.
   */
  verificar(documentoId: number, observaciones?: string): Observable<DocumentoExpedienteDTO> {
    return this.http.post<DocumentoExpedienteDTO>(
      `${this.apiUrl}/${documentoId}/verificar`,
      observaciones ?? null
    );
  }

  /**
   * Rechaza un documento indicando el motivo.
   * El familiar deberá presentar nuevamente el documento.
   */
  rechazar(documentoId: number, motivo: string): Observable<DocumentoExpedienteDTO> {
    const dto: Partial<RechazarDocumentoDTO> = { motivo };
    return this.http.post<DocumentoExpedienteDTO>(
      `${this.apiUrl}/${documentoId}/rechazar`,
      dto
    );
  }

  // ═══════════════════════════════════════════════════════════
  // DESCARGA / ELIMINACIÓN
  // ═══════════════════════════════════════════════════════════

  /**
   * Descarga el archivo de un documento como Blob
   */
  descargar(documentoId: number): Observable<Blob> {
    return this.http.get(
      `${this.apiUrl}/${documentoId}/descargar`,
      { responseType: 'blob' }
    );
  }

  /**
   * Elimina un documento.
   * Solo permitido si Estado == PendienteVerificacion o Rechazado.
   */
  eliminar(documentoId: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${documentoId}`);
  }

  // ═══════════════════════════════════════════════════════════
  // HELPERS
  // ═══════════════════════════════════════════════════════════

  /**
   * Descarga un Blob como archivo en el navegador
   */
  descargarArchivo(blob: Blob, nombreArchivo: string): void {
    const url = window.URL.createObjectURL(blob);
    const link = document.createElement('a');
    link.href = url;
    link.download = nombreArchivo;
    link.click();
    window.URL.revokeObjectURL(url);
  }

  /**
   * Valida formato del archivo antes de subir
   * Formatos permitidos: .pdf, .jpg, .jpeg, .png
   */
  validarFormato(archivo: File): boolean {
    const extensionesPermitidas = ['.pdf', '.jpg', '.jpeg', '.png'];
    const extension = '.' + archivo.name.split('.').pop()?.toLowerCase();
    return extensionesPermitidas.includes(extension);
  }

  /**
   * Valida tamaño del archivo (máximo 5MB)
   */
  validarTamanio(archivo: File): boolean {
    const maxBytes = 5 * 1024 * 1024;
    return archivo.size <= maxBytes;
  }

  /**
   * Formatea bytes a texto legible (KB, MB)
   */
  formatearTamanio(bytes: number): string {
    if (bytes === 0) return '0 Bytes';
    const k = 1024;
    const sizes = ['Bytes', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(k));
    return Math.round((bytes / Math.pow(k, i)) * 100) / 100 + ' ' + sizes[i];
  }

  /**
   * Obtiene el color del badge según el estado del documento
   */
  obtenerColorEstado(estado: EstadoDocumentoExpediente): string {
    switch (estado) {
      case EstadoDocumentoExpediente.Verificado:
        return 'bg-green-100 text-green-800 border-green-300';
      case EstadoDocumentoExpediente.Rechazado:
        return 'bg-red-100 text-red-800 border-red-300';
      case EstadoDocumentoExpediente.PendienteVerificacion:
      default:
        return 'bg-yellow-100 text-yellow-800 border-yellow-300';
    }
  }

  /**
   * Obtiene el icono según el estado del documento
   */
  obtenerIconoEstado(estado: EstadoDocumentoExpediente): string {
    switch (estado) {
      case EstadoDocumentoExpediente.Verificado:
        return 'check-circle';
      case EstadoDocumentoExpediente.Rechazado:
        return 'x-circle';
      case EstadoDocumentoExpediente.PendienteVerificacion:
      default:
        return 'clock';
    }
  }

  /**
   * Obtiene la descripción del tipo de documento
   */
  obtenerDescripcionTipo(tipo: TipoDocumentoExpediente): string {
    switch (tipo) {
      case TipoDocumentoExpediente.DNI_Familiar:
        return 'DNI del Familiar';
      case TipoDocumentoExpediente.DNI_Fallecido:
        return 'DNI del Fallecido';
      case TipoDocumentoExpediente.CertificadoDefuncion:
        return 'Certificado de Defunción (SINADEF)';
      case TipoDocumentoExpediente.OficioLegal:
        return 'Oficio Legal (PNP/Fiscal)';
      case TipoDocumentoExpediente.ActaLevantamiento:
        return 'Acta de Levantamiento';
      case TipoDocumentoExpediente.Otro:
        return 'Documento Adicional';
      default:
        return 'Documento';
    }
  }
}
